### DISCUSSION TITLE
How do I completely cleanup SQLAlchemy at the end of a test?

### DISCUSSION THREAD

**Discussion by mcarans** [sqlalchemy/sqlalchemy#11709] [Usage Questions] [answered]
I have been struggling with a simple test that passes when run standalone or first, but fails with a "no such table" error when it is run after other tests that use SQLAlchemy. I've tried all sorts of cleaning but nothing seems to completely clean things up so that this test will pass.

I tried cleaning at the end of the tests with some combination of these:

```
Base.metadata.drop_all()
Base.metadata.clear()
session.close()
engine.dispose()

```

I've also tried a scoped_session that I reset but it didn't help.

Is there a way to completely clear all SQLAlchemy state at the end of a test?

EDIT: I have managed to replicate the issue in a new [repo](https://github.com/mcarans/sqlalchemyissue) with cut down code - see this [test failure](https://github.com/mcarans/sqlalchemyissue/actions/runs/10313307826/job/28549861333).

What is particularly strange is it only seems to fail like this in GitHub Actions. Running `hatch test` or pytest in a virtualenv locally works.

**Comment by mcarans** [comment]
I realised that I should not do
```
Base.metadata.drop_all()
Base.metadata.clear()
```

**Comment by mcarans** [comment]
I spoke too soon. A different test that uses SQLAlchemy reflection sometimes fails (depending upon the order of tests) saying it cannot find attribute table1 (where table1 is a reflected class). So I still need to find out what state is held between tests and clear it somehow.

**Comment by CaselIT** [comment]
Hi,

Please note that without examples there isn't much to suggest.

**Reply by mcarans** [reply]
Fair point. If I can find the time, I'll try to make cut down standalone example repo that demonstrates the problem.

**Reply by mcarans** [reply]
I have managed to replicate it in a new [repo](https://github.com/mcarans/sqlalchemyissue) with cut down code - see this [test failure](https://github.com/mcarans/sqlalchemyissue/actions/runs/10313307826/job/28549861333).

What is particularly strange is it only seems to fail like this in GitHub Actions. Running `hatch test`  or pytest in a virtualenv locally works.

**Reply by CaselIT** [reply, marked as answer]
Well I would start by removing the pass on oserrors. The fact that there are no tables suggests that there was an error there

**Reply by mcarans** [reply]
Thanks for your help!. I changed my code so that completely different filenames are used for the two tests so that removing the SQLite db file between tests isn't needed. That fixed it. Perhaps the file was locked and so couldn't be deleted or changed?

**Reply by CaselIT** [reply]
maybe. overall it's never a great idea to completely ignore exceptions
